kie-server-parent/kie-server-controller/kie-server-controller-openshift/src/test/java/org/kie/server/controller/openshift/storage/OpenShiftServerTemplateStorageTest.java

```diff
@@ -0,0 +1,208 @@
+        KieServerStateOpenShiftRepository repo = new KieServerStateOpenShiftRepository() {
+
+            @Override
+            public OpenShiftClient createOpenShiftClient() {
+                return client;
+            }
+
+            @Override
+            public KubernetesClient createKubernetesClient() {
+                return client;
+            }
+
+            @Override
+            public boolean isKieServerReady() {
+                return true;
+            }
+            
+            @Override
+            public boolean isDCStable(DeploymentConfig dc) {
+                return true;
+            }
+        };
+
+        templateStorage = new OpenShiftServerTemplateStorage(repo);
+        createServerTemplateWithContainer();
+    }
+
+    @Override
+    protected void checkTemplateInstances(ServerTemplate fromStorage) {
+        Collection<ServerInstanceKey> instances = fromStorage.getServerInstanceKeys();
+        assertNotNull(instances);
+
+        //OpenShift based ServerTemplate contains one and only one ServerInstance
+        assertEquals(1, instances.size());
+    }
+
+    @Override
+    protected boolean isRequired() {
+        return false;
+    }
+
+    @Test
+    public void testLoadKeysWithCache() {
+        OpenShiftServerTemplateStorage ocpStorage = (OpenShiftServerTemplateStorage) templateStorage;
+        ocpStorage.store(serverTemplate);
+        assertEquals(1, ocpStorage.loadKeys().size());
+
+        // Removed directly from KieServerStateOpenShiftRepository
+        ocpStorage.repo.delete(serverTemplate.getId());
+
+        // Still in the cache
+        assertEquals(1, ocpStorage.loadKeys().size());
+    }
+
+    @Test
+    public void testLoadKeysWithCacheExpire() {
+        OpenShiftServerTemplateStorage ocpStorage = (OpenShiftServerTemplateStorage) templateStorage;
+
+        // Set TTL to 10 milliseconds
+        ocpStorage.cacheTTL = 10;
+
+        ocpStorage.store(serverTemplate);
+        assertEquals(1, ocpStorage.loadKeys().size());
+
+        // Removed directly from KieServerStateOpenShiftRepository
+        ocpStorage.repo.delete(serverTemplate.getId());
+
+        try {
+            Thread.sleep(500);
+        } catch (Exception e) {
+        }
+
+        // Cache should be expired
+        assertEquals(0, ocpStorage.loadKeys().size());
+    }
+
+    @Test
+    public void testLoadWithCache() {
+        OpenShiftServerTemplateStorage ocpStorage = (OpenShiftServerTemplateStorage) templateStorage;
+        ocpStorage.store(serverTemplate);
+        assertEquals(1, ocpStorage.load().size());
+
+        // Removed directly from KieServerStateOpenShiftRepository
+        ocpStorage.repo.delete(serverTemplate.getId());
+
+        // Still in the cache
+        assertEquals(1, ocpStorage.load().size());
+    }
+    
+    @Override
+    @Test
+    public void testStoreLoadAndUpdateServerTemplate() {
+        createDummyDC(serverTemplate.getId());
```

Comment:
What is the DC used for here?

Reply:
The reason for adding a dummy DC is to simulate the attached CM scenario, meaning that any attempt to delete a CM/ServerTemplate which has already been used by KieServer Pod (A DC with same name exists) will be blocked by throwing an error.

A gap in the original test method implementation has been fixed by a new test method. 

https://github.com/rhtevan/droolsjbpm-integration/blob/fd4635e8e9b2263458db3339d7ea0c94b7ae87b6/kie-server-parent/kie-server-controller/kie-server-controller-openshift/src/test/java/org/kie/server/controller/openshift/storage/OpenShiftServerTemplateStorageTest.java#L167-L177